### ISSUE TITLE
How do we prevent broadly used packages from bit-rotting?

### ISSUE THREAD

**Issue by dmose** [dmose/mc-nodejs-docs#12]
@loganfsmyth points out that in some situations, it can be a ton of work to upgrade broadly used packages, causing it to lag:

> Would this require anyone updating a broadly-used package to upgrade every codebase that uses it, or can you vendor in a new major version and upgrade your own codebase to use it, while leaving it up to other codebases to decide when to upgrade theirs? I'm not familiar with how this is handled for other vendored dependencies. If you have to upgrade everything, how do we stop ourselves from getting into a place where upgrading is essentially too much work, given that node's normal setup makes upgrading an easy per-project process.

CC @mshal, @ncalexan

**Comment by dmose** [comment]
@globau, do you have any thoughts here?

**Comment by ncalexan** [comment]
> @loganfsmyth points out that in some situations, it can be a ton of work to upgrade broadly used packages, causing it to lag:
> 
> > Would this require anyone updating a broadly-used package to upgrade every codebase that uses it, or can you vendor in a new major version and upgrade your own codebase to use it, while leaving it up to other codebases to decide when to upgrade theirs? I'm not familiar with how this is handled for other vendored dependencies. If you have to upgrade everything, how do we stop ourselves from getting into a place where upgrading is essentially too much work, given that node's normal setup makes upgrading an easy per-project process.
> 
> CC @mshal, @ncalexan

There's no silver bullet here.  Yarn, and perhaps NPM, "lift" dependencies as far up the `node_modules` hierarchy as possible until a conflict is determined.  I won't draw a diagram; there are lots of these floating around.  That means that a tree can contain multiple versions of a package and the Node.js resolution algorithm will find them appropriately.

So if we read "codebase" as "subdirectory in m-c", then as long as that subdirectory itself has a `package.json` then I expect the Node.js resolution algorithm will accommodate multiple incompatible versions.  (My hazy memory here is that the details are unsavoury: one ends up with `node_modules` littered throughout the source tree.  Perhaps this has improved or perhaps there are workarounds I am unaware of.)

**Comment by dmose** [comment]
>  (My hazy memory here is that the details are unsavoury: one ends up with node_modules littered throughout the source tree. Perhaps this has improved or perhaps there are workarounds I am unaware of.)

You are correct, there are various node_modules littered throughout the source tree. My hope is that once this project finishes, people will start hoisting at least some of their stuff to the top-level, if for no other reason than to make it be directly accessible by the build system.

Whether the other node_modules ever actually go away is not totally clear to me, but this project is at least necessary if not sufficient.
